File: 0061-rotate-list/0061-rotate-list.cpp
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || k == 0) return head;

        // First, find the length of the list and the last node
        ListNode* last = head;
        int length = 1;
        while (last->next) {
            last = last->next;
            length++;
        }

        // Connect the last node to the head to make it a circular list
        last->next = head;

        // Find the new tail position
        k = k % length; // To handle cases where k is greater than length
        int stepsToNewHead = length - k;

        ListNode* newTail = head;
        for (int i = 1; i < stepsToNewHead; i++) {
            newTail = newTail->next;
        }

        // Set the new head and break the circular link
        ListNode* newHead = newTail->next;
        newTail->next = nullptr;

        return newHead;
    }
};

File: 0061-rotate-list/0061-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        #Solution 2 Optimal
        n = 0
        crrNode = head
        oldtail = crrNode
        while crrNode is not None:
            n+=1
            oldtail = crrNode
            crrNode = crrNode.next
        oldhead = head
        
        if n<2:
            return head

        k=k%n
        if k==0:
            return head

        crrNode = head
        for _ in range(n-k-1):
            crrNode = crrNode.next
        newtail = crrNode
        newhead = newtail.next
        oldtail.next = oldhead
        newtail.next = None

        return newhead